jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -65,29 +75,74 @@
     private final AtomicInteger pending = new AtomicInteger();
+    private final Strategy strategy;
+    private final ThreadLocal<Entry> cache;
+    private final AtomicInteger next;
     private volatile boolean closed;
     private volatile int maxMultiplex = 1;
     private volatile int maxUsageCount = -1;
 
+    public enum Strategy
+    {
+        /**
+         * The Linear strategy looks for an entry always starting from the first entry.
+         * It will favour the early entries in the pool, but may contend on them more.
+         */
+        LINEAR,
+
+        /**
+         * The Random strategy looks for an entry by iterating from a random starting
+         * index.  No entries are favoured and contention is reduced.
+         */
+        RANDOM,
+
+        /**
+         * The Thread ID strategy uses the {@link Thread#getId()} of the current thread
+         * to select a starting point for an entry search.  Whilst not as performant as
+         * using the {@link ThreadLocal} cache, it may be suitable when the pool is substantially smaller
+         * than the number of available threads.
+         * No entries are favoured and contention is reduced.
+         */
+        THREAD_ID,
+
+        /**
+         * The Round Robin strategy looks for an entry by iterating from a starting point
+         * that is incremented on every search. This gives similar results to the
+         * random strategy but with more predictable behaviour.
+         * No entries are favoured and contention is reduced.
+         */
+        ROUND_ROBIN,
+    }
+
     /**
-     * Construct a Pool with the specified thread-local cache size.
+     * Construct a Pool with a specified lookup strategy and no
+     * {@link ThreadLocal} cache.
      *
+     * @param strategy The strategy to used for looking up entries.
      * @param maxEntries the maximum amount of entries that the pool will accept.
-     * @param cacheSize the thread-local cache size. A value less than 1 means the cache is disabled.
      */
-    public Pool(int maxEntries, int cacheSize)
+    public Pool(Strategy strategy, int maxEntries)
+    {
+        this(strategy, maxEntries, false);
+    }
+
+    /**
+     * Construct a Pool with the specified thread-local cache size and
+     * an optional {@link ThreadLocal} cache.
+     * @param strategy The strategy to used for looking up entries.
+     * @param maxEntries the maximum amount of entries that the pool will accept.
+     * @param cache True if a {@link ThreadLocal} cache should be used to try the most recently released entry.
+     */
+    public Pool(Strategy strategy, int maxEntries, boolean cache)
     {
         this.maxEntries = maxEntries;
-        this.cacheSize = cacheSize;
-        if (cacheSize > 0)
-            this.cache = ThreadLocal.withInitial(() -> new ArrayList<Entry>(cacheSize));
-        else
-            this.cache = null;
+        this.strategy = strategy;
+        this.cache = cache ? new ThreadLocal() : null;
```

Comment:
`new ThreadLocal<>()`?

Reply:
Don't need the threadlocal if not caching